cirq-google/cirq_google/named_topologies.py

```diff
@@ -0,0 +1,259 @@
+from functools import lru_cache
+from typing import Dict, List, Tuple, Any, Sequence
+
+import cirq
+import networkx as nx
+from matplotlib import pyplot as plt
+
+
+def cache(user_function):
+    """Unbounded cache.
+
+    Available as functools.cache in Python 3.9+
+    """
+    return lru_cache(maxsize=None)(user_function)
+
+
+def dataclass_json_dict(obj, namespace: str = None):
+    return cirq.obj_to_dict_helper(
+        obj, [f.name for f in dataclasses.fields(obj)], namespace=namespace
+    )
+
+
+class NamedTopology(metaclass=abc.ABCMeta):
+    """A named topology."""
+
+    name = NotImplemented
+    """A name that uniquely identifies this topology."""
+
+    n_nodes = NotImplemented
+    """The number of nodes in the topology."""
+
+    graph = NotImplemented
+    """A networkx graph representation of the topology."""
+
+
+def _node_and_coordinates(nodes):
+    for node in nodes:
+        if isinstance(node, cirq.GridQubit):
+            yield node, (node.row, node.col)
+        else:
+            x, y = node
+            yield node, (x, y)
+
+
+def draw_gridlike(
+    graph: nx.Graph, ax=None, cartesian: bool = True, **kwargs
+) -> Dict[Any, Tuple[int, int]]:
+    """Draw a Grid-like graph.
+
+    Args:
+        graph: A NetworkX graph whose nodes are (row, column) coordinates.
+        ax: Optional matplotlib axis to use for drawing.
+        cartesian: If True, directly position as (row, column); otherwise,
+            rotate 45 degrees to accommodate google-style diagonal grids.
+        kwargs: Additional arguments to pass to `nx.draw_networkx`.
+
+    Returns:
+        A positions dictionary mapping nodes to (x, y) coordinates suitable for future calls
+        to NetworkX plotting functionality.
+    """
+    if ax is None:
+        ax = plt.gca()  # coverage: ignore
+
+    if cartesian:
+        pos = {node: (y, -x) for node, (x, y) in _node_and_coordinates(graph.nodes)}
+    else:
+        pos = {node: (x + y, y - x) for node, (x, y) in _node_and_coordinates(graph.nodes)}
+
+    nx.draw_networkx(graph, pos=pos, ax=ax, **kwargs)
+    ax.axis('equal')
+    return pos
+
+
+@dataclass(frozen=True)
+class LineTopology(NamedTopology):
+    """A 1D linear topology.
+
+    Node indices are contiguous integers starting from 0 with edges between
+    adjacent integers.
+
+    Args:
+        n_nodes: The number of nodes in a line.
+    """
+
+    n_nodes: int
+
+    # https://github.com/python/mypy/issues/1362
+    @property  # type: ignore
+    @cache
+    def name(self) -> str:
+        """The name of this topology: {n_nodes}-line"""
+        return f'{self.n_nodes}-line'
+
+    @property  # type: ignore
+    @cache
+    def graph(self) -> nx.Graph:
+        """The graph of this topology.
+
+        Node indices are contiguous integers starting from 0 with edges between
+        adjacent integers.
+        """
+        return nx.from_edgelist(
+            [(i1, i2) for i1, i2 in zip(range(self.n_nodes), range(1, self.n_nodes))]
```

Comment:
Hmm. It won't have two nodes. It would have zero. Which isn't right either :) 

for n_nodes = 1, the first range will be [0] and the second would be []. Zip stops after either iterator is exhausted so there would be no edges. We rely on this zip behavior or else the first range would have to stop 1 earlier.

I'll put in a check that `n_nodes > 1` it doesn't make much sense as a line otherwise.